File: frontend/src/stores/notification.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import api from '../composables/useApi'

export interface Notification {
  id: string
  type: string
  actor: {
    id: string
    username: string
    avatar_url: string | null
  }
  post_id: string | null
  comment_id: string | null
  is_read: boolean
  created_at: string
}

export const useNotificationStore = defineStore('notification', () => {
  const notifications = ref<Notification[]>([])
  const unreadCount = ref(0)

  const hasUnread = computed(() => unreadCount.value > 0)

  async function fetchNotifications() {
    const response = await api.get('/notifications')
    const data = response.data
    notifications.value = data.notifications
    unreadCount.value = data.unread_count
  }

  async function markAsRead(id: string) {
    await api.patch(`/notifications/${id}/read`)
    const notification = notifications.value.find((n) => n.id === id)
    if (notification && !notification.is_read) {
      notification.is_read = true
      unreadCount.value = Math.max(0, unreadCount.value - 1)
    }
  }

  async function markAllAsRead() {
    await api.patch('/notifications/read-all')
    notifications.value.forEach((n) => {
      n.is_read = true
    })
    unreadCount.value = 0
  }

  function addNotification(notification: Notification) {
    notifications.value.unshift(notification)
    if (!notification.is_read) {
      unreadCount.value++
    }
  }

  return {
    notifications,
    unreadCount,
    hasUnread,
    fetchNotifications,
    markAsRead,
    markAllAsRead,
    addNotification,
  }
})

File: backend/src/routes/notification.rs
use axum::{Router, Json, routing::{get, patch}, extract::{Path, Query, State}};
use std::sync::Arc;

use crate::models::AppState;
use crate::extractors::AuthenticatedUser;
use crate::services::NotificationService;
use crate::models::response::PaginationQuery;
use crate::error::Result;

pub fn notification_routes() -> Router<Arc<AppState>> {
    Router::new()
        .route("/", get(list_notifications))
        .route("/read-all", patch(mark_all_read))
        .route("/{id}/read", patch(mark_read))
}

async fn list_notifications(
    State(state): State<Arc<AppState>>,
    AuthenticatedUser(user): AuthenticatedUser,
    Query(query): Query<PaginationQuery>,
) -> Result<Json<crate::models::notification::NotificationListResponse>> {
    let notification_service = NotificationService::new(&state.db);
    notification_service.list_notifications(&user.id, query.page(), query.limit(), false).await.map(Json)
}

async fn mark_read(
    State(state): State<Arc<AppState>>,
    AuthenticatedUser(user): AuthenticatedUser,
    Path(id): Path<String>,
) -> Result<Json<crate::models::notification::NotificationWithActor>> {
    let notification_service = NotificationService::new(&state.db);
    notification_service.mark_read(&user.id, &id).await?;

    let notification_repo = crate::repositories::NotificationRepository::new(&state.db);
    let notifications = notification_repo.list(&user.id, 1, 1, false).await?.0;

    Ok(Json(notifications.into_iter().find(|n| n.id == id)
        .unwrap_or(crate::models::notification::NotificationWithActor {
            id: id.clone(),
            notification_type: "read".to_string(),
            actor: crate::models::user::UserPublic {
                id: user.id.clone(),
                username: user.username.clone(),
                avatar_url: None,
                bio: None,
                is_admin: user.is_admin,
                created_at: chrono::Utc::now().to_rfc3339(),
            },
            post_id: None,
            comment_id: None,
            is_read: true,
            created_at: chrono::Utc::now().to_rfc3339(),
        })))
}

async fn mark_all_read(
    State(state): State<Arc<AppState>>,
    AuthenticatedUser(user): AuthenticatedUser,
) -> Result<Json<crate::models::response::MessageResponse>> {
    let notification_service = NotificationService::new(&state.db);
    notification_service.mark_all_read(&user.id).await?;

    Ok(Json(crate::models::response::MessageResponse {
        message: "All notifications marked as read".to_string()
    }))
}
